reana_workflow_controller/rest.py

```diff
@@ -1526,3 +1528,57 @@ def _get_current_job_progress(workflow_id):
                 'prettified_cmd': job.prettified_cmd,
                 'current_job_name': job.name}
     return current_job_commands
+
+
+def _delete_workflow(workflow, all_runs=False, hard_delete=False):
+    """Delete workflow."""
+    if workflow.status in [WorkflowStatus.created,
+                           WorkflowStatus.finished,
+                           # WorkflowStatus.stopped
+                           WorkflowStatus.failed]:
+        to_be_deleted = [workflow]
+        if all_runs:
+            to_be_deleted += Session.query(Workflow).\
+                filter_by(name=workflow.name).all()
+        for workflow in to_be_deleted:
+            if hard_delete:
+                _hard_delete_workflow(workflow)
+            else:
+                _soft_delete_workflow(workflow)
+
+        return jsonify({'message': 'Workflow successfully deleted',
+                        'workflow_id': workflow.id_,
+                        'workflow_name': _get_workflow_name(workflow),
+                        'status': workflow.status.name,
+                        'user': str(workflow.owner_id)}), 200
+    elif workflow.status == WorkflowStatus.deleted:
+        raise WorkflowDeletionError(
+            'Workflow {0}.{1} has already been deleted.'.
+            format(
+                workflow.name,
+                workflow.run_number))
+    else:
+        raise WorkflowDeletionError(
+            'Workflow {0}.{1} cannot be deleted as it'
+            ' is currently running. If you are sure '
+            'you want to delete it provide the '
+            '--force flag.'.
+            format(
+                workflow.name,
+                workflow.run_number))
+
+
+def _hard_delete_workflow(workflow):
+    """Remove workflow row from database."""
+    remove_workflow_workspace(workflow.get_workspace())
+    Session.query(Workflow).filter_by(id_=workflow.id_).delete()
+    Session.commit()
+
+
+def _soft_delete_workflow(workflow):
+    """Mark workflow as deleted."""
+    remove_workflow_workspace(workflow.get_workspace())
```

Comment:
We could consider leaving the workspace in place for a short duration of time (like 1 day) and removing it afterwards, to allow users to download files from it.

Reply:
I would go for now with the simpler approach of removing it straightaway since we would have to put in place more infrastructure to perform cron jobs.